src/coreclr/src/System.Private.CoreLib/src/Internal/Runtime/InteropServices/FunctionPointerLoader.cs

```diff
@@ -0,0 +1,153 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
+
+namespace Internal.Runtime.InteropServices
+{
+    public static class FunctionPointerLoader
+    {
+        private static readonly Dictionary<IntPtr, Delegate> s_delegates = new Dictionary<IntPtr, Delegate>();
+
+        private static string MarshalToString(IntPtr arg, string argName)
+        {
+            if (arg == IntPtr.Zero)
+            {
+                throw new ArgumentNullException(argName);
+            }
+
+#if TARGET_WINDOWS
+            string? result = Marshal.PtrToStringUni(arg);
+#else
+            string? result = Marshal.PtrToStringUTF8(arg);
+#endif
+
+            if (result == null)
+            {
+                throw new ArgumentNullException(argName);
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Native hosting entry point for creating a native delegate
+        /// </summary>
+        /// <param name="typeNameNative">Assembly qualified type name</param>
+        /// <param name="methodNameNative">Public static method name compatible with delegateType</param>
+        /// <param name="delegateTypeNative">Assembly qualified delegate type name</param>
+        /// <param name="loadContext">Extensibility parameter (currently unused)</param>
+        /// <param name="reserved">Extensibility parameter (currently unused)</param>
+        /// <param name="functionHandle">Pointer where to store the function pointer result</param>
+        [UnmanagedCallersOnly]
+        public static unsafe int GetFunctionPointer(IntPtr typeNameNative,
```

Comment:
I would add this to the `ComponentActivator` and not create a new class:
* No need to duplicate code (the `MarshalToString` method for example)
* We can share the `s_delegates`, no reason to have two dictionaries for basically the same thing
* More options for sharing code - the method body here is very similar to the `LoadAssemblyAndGetFunctionPointer` so ideally we would rewrite it such that there's little to no duplication

Reply:
I do see that the name `ComponentActivator` will then become somewhat weird. We can discuss changing that if we feel it's really bad. Even though it's a public class, it's not part of any public contract - `System.Private.CoreLib` has quite a few public APIs which are not truly public (nobody compiles directly against S.P.CoreLib except the rest of the framework). And the only other place where it's used is `hostpolicy` which always ships with the runtime, so there will never be a case where older/newer `hostpolicy` would interact with newer/older runtime.

Reply:
My thinking was that the assembly path parameter made it fundamentally different, along with the fact that the future extensibility parameters were also likely to be different from each other. There might be a problem with the design - how is `get_function_pointer` ever going to support an isolated ALC if it doesn't have a way to specify the assembly path?

Reply:
My view is that `LoadAssemblyAndGetFunctionPointer` is basically a concatenation of two relatively distinct parts:
* Load assembly
* Get function pointer

You're now adding the second part as a separately accessible piece (with limitations). The goal of "Get function pointer" should not be to worry about loading assemblies or creating load contexts. That's the job of the first part "Load assembly". Long term I would like to introduce a separately callable function which does just that part - load assembly with some options as to where (default ALC, new isolated ALC, existing secondary ALC, ...). That's why we're adding the `loadContext` parameter to `get_function_pointer` - basically the future "load assembly" would return the value to be passed to `get_function_pointer`.

I don't see a problem with the specific runtime helpers (`LoadAssemblyAndGetFunctionPointer` and now `GetFunctionPointer`) having different parameters and somewhat different behavior while living on the same class. To me they have a lot more in common as can be seen by the implementation which is very similar.

If you think that having a separate class would be better, I don't have a problem with that (I personally don't see the point, but that's just me) - what I do want to see is sharing the code which should work the same.